tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/arena/PtrUtil.scala

```diff
@@ -11,6 +14,38 @@ import at.forsyte.apalache.tla.types.tla
  */
 object PtrUtil {
 
+  type CachingHeuristic = TlaEx => Boolean
+
+  object Heuristics {
+
+    // for debugging; using this is equivalent to disabling caching
+    def alwaysFalse: CachingHeuristic = _ => false
+
+    def notAName: CachingHeuristic = {
+      case _: NameEx => false
+      case _         => true
+    }
+
+    def notConjunctionOfNames: CachingHeuristic = {
+      case _: NameEx                          => false
+      case OperEx(TlaBoolOper.and, args @ _*) => args.exists(notConjunctionOfNames)
+      case _                                  => true
+    }
```

Comment:
This does not seem to be a complex expression to me. I was expecting something like a threshold over the expression size.

Reply:
I mean, these were really just prototypes meant to be slightly more complex than no-ops, not the final versions. Obviously the actual heuristic will ultimately need some more planning.